src/ArborX_HDBSCAN.hpp

```diff
@@ -0,0 +1,163 @@
+  Kokkos::View<float *, MemorySpace> core_distances(
+      NoInit("ArborX::HDBSCAN::core_distances"), n);
+  constexpr auto inf = KokkosExt::ArithmeticTraits::infinity<float>::value;
+  Kokkos::deep_copy(exec_space, core_distances, -inf);
+  bvh.query(exec_space,
+            Details::NearestK<Primitives>{primitives, core_min_size},
+            ArborX::Details::MaxDistance<Primitives, decltype(core_distances)>{
+                primitives, core_distances});
+
+  Kokkos::Profiling::popRegion();
+  Kokkos::Profiling::pushRegion("ArborX::HDBSCAN::boruvka");
+
+  Kokkos::View<int *, MemorySpace> cached_closest_neighbors(
+      NoInit("ArborX::HDBSCAN::neighbors"), n);
+  ArborX::iota(exec_space, cached_closest_neighbors);
+
+  Kokkos::View<int *, MemorySpace> labels(NoInit("ArborX::HDBSCAN::labels"), n);
+  ArborX::iota(exec_space, labels);
+
+  Kokkos::View<Kokkos::pair<int, int> *, MemorySpace> components_out_edges(
+      NoInit("ArborX::HDBSCAN::components_out_edges"), n);
+
+  Kokkos::View<Kokkos::pair<int, int> *, MemorySpace> mst_edges(
+      NoInit("ArborX::HDBSCAN::mst_edges"), n - 1);
+
+  // Compute parents
+  Kokkos::View<int *, MemorySpace> bvh_parents(
+      NoInit("ArborX::HDBSCAN::parents"), 2 * n - 1);
+  Details::findParents(exec_space, bvh, bvh_parents);
+
+  using Predicates = Details::NearestK<Primitives>;
+  Predicates predicates{primitives, 1};
+
+  Kokkos::Profiling::pushRegion("ArborX::HDBSCAN::compute_permutation");
+  // Permute predicates once for all Boruvka iterations
+  using DeviceType = Kokkos::Device<ExecutionSpace, MemorySpace>;
+  auto permute =
+      Details::BatchedQueries<DeviceType>::sortQueriesAlongZOrderCurve(
+          exec_space, static_cast<Box>(bvh.bounds()), predicates);
+  Kokkos::Profiling::popRegion();
+
+  using PermutedPredicates =
+      Details::PermutedData<Predicates, decltype(permute)>;
+  PermutedPredicates permuted_predicates{predicates, permute};
+
+  // Boruvka loop
+  int it = 0;
+  int num_components = n;
+  float boruvka_time = 0.f;
+  while (num_components > 1)
+  {
+    Kokkos::Timer timer_boruvka_iteration;
+    if (verbose)
+    {
+      printf("#%4d: [%10d]", ++it, num_components);
+      exec_space.fence();
+      timer_boruvka_iteration.reset();
+    }
+
+    Details::determineComponentEdges(exec_space, permuted_predicates, bvh,
+                                     bvh_parents, labels, core_distances,
+                                     components_out_edges);
+
+    num_components = Details::updateLabels(
+        exec_space, num_components, components_out_edges, labels, mst_edges);
+
+    exec_space.fence();
+    if (verbose)
+    {
+      exec_space.fence();
+      auto time_iteration = timer_boruvka_iteration.seconds();
+      boruvka_time += time_iteration;
+      printf("  time : %10.3f\n", time_iteration);
+    }
+  }
+  if (verbose)
+  {
+    printf("#%4d: [%10d]\n", ++it, num_components);
+    printf("Total Boruvka time  : %10.3f\n", boruvka_time);
+  }
+
```

Comment:
What do you consider to be the product of the Boruvka algorithm?  `component_out_edges`?   Do you use the `labels` in the next stages?  Do you need to recompute the weights?

Reply:
The result is an MST, a collection of undirected edges. Depending on usefulness, we may want to further restrict this. For example, we could require `(i, j, d)` with `i <= j` and `d = mutual_reachability_distance(i, j)`. In fact, storing the latter one would allow us to avoid accessing primitives and core distances later.